docs/extensions.rst

```diff
@@ -4,49 +4,215 @@
+``putup --dry-run``.
+
+
+What are Extensions?
+====================
+
+From the standpoint of PyScaffold, extensions are composed by a group of
+functions that can be used to inject actions at arbitrary positions in the
+aforementioned list. Furthermore, extensions can also remove actions.
 
 
 Creating an Extension
 =====================
 
-Extensions should implement a regular python function that receives as a single
-argument a :class:`~pyscaffold.api.Scaffold` object.
-This object is a representation of all the actions that will be performed by
-the ``putup`` command, and contain the following properties:
-
-:attr:`~pyscaffold.api.Scaffold.options` (*dict*)
-  all PyScaffold options, including the ones parsed from command line;
-:attr:`~pyscaffold.api.Scaffold.before_generate` (*list*)
-  functions that will be executed **before** the generation of files;
-:attr:`~pyscaffold.api.Scaffold.after_generate` (*list*)
-  functions that will be executed **after** the generation of files;
-:attr:`~pyscaffold.api.Scaffold.structure` (*dict*)
-  directory tree representation as a (possibly nested) dictionary.
-  The keys indicate each part of the path for the generated file,
-  while the value indicate its contents.
-:attr:`~pyscaffold.api.Scaffold.changed_structure` (*dict*)
-  that is very similar to :attr:`~pyscaffold.api.Scaffold.structure`,
-  but only stores files that actually change
-  (this attribute is only assigned after project creation).
-
-Additionally, the following methods are also available:
-
-:obj:`~pyscaffold.api.Scaffold.merge_structure`
-  deep merge the dictionary argument with the current representation of the
-  to-be-generated directory tree.
-:obj:`~pyscaffold.api.Scaffold.ensure_file`
-  ensure a single file exists in the current representation of the tree,
-  with the provided content, automatically creating the parent directories.
-:obj:`~pyscaffold.api.Scaffold.reject_file`
-  remove a file from the tree representation if the entire path exists.
-
-The ``project`` and  ``package`` options can be used in order to ensure the
-correct location of the files.
-
-The following example illustrates a simple extension implementation:
+In order to create an extension it is necessary to implement a regular python
+function that:
+
+- receives a list of actions as first argument,
+- receives a :mod:`namespace object <pyscaffold.api.helpers>` containing some
+  helper methods as second argument,
+- registers custom actions that will be called later and
+- returns a modified version of the list of actions.
+
+The following pseudo-code highlights a typical definition of such kind of
+functions:
+
+.. code-block:: python
+
+    def extend_scaffold(actions, helpers):
+        """Register custom actions that apply extension behavior."""
+
+        # ... define `start_hook`, `finish_hook` and `define_awesome_files`
+        #     functions
+
+        actions = helpers.register(actions, start_hook, before='create_structure')
```

Comment:
Great explanation! Should it mean rather `before='define_structure'` here? Wouldn't that be more aligned with the more detailed explanation below?

Reply:
Do you think it is easier to understand? How about `after='create_structure'` should I keep it or change to `'define_structure'` as well?

Reply:
Ahh.. maybe that's just my mistake. I was comparing it to the later defined 
```python
helpers.register(actions, hook1, before='define_structure')
helpers.register(actions, hook2, after='pyscaffold.structure:create_structure')
```
and was confused. At that point I mixed something up. So it's okay from my point.